beacon-chain/db/kv/blobs.go

```diff
@@ -1,58 +1,93 @@
-			return err
-		}
-		return tx.Bucket(blobsAgesBucket).Delete(root[:])
-	})
-}
-
-// SaveBlobsSidecar saves the blobs for a given epoch in the sidecar bucket.
-func (s *Store) SaveBlobsSidecar(ctx context.Context, blob *ethpb.BlobsSidecar) error {
+const blobSidecarKeyLength = 48 // slot_to_rotating_buffer(blob.slot) ++ blob.slot ++ blob.block_root
+
+// SaveBlobsSidecar saves the blobs for a given epoch in the sidecar bucket. When we receive a blob:
+// 1. Convert slot using a modulo operator to [0, maxSlots] where maxSlots = MAX_BLOB_EPOCHS*SLOTS_PER_EPOCH
+// 2. Compute key for blob as bytes(slot_to_rotating_buffer(blob.slot)) ++ bytes(blob.slot) ++ blob.block_root
+// 3. Begin the save algorithm:  If the incoming blob has a slot bigger than the last saved slot at the spot
+//    in the rotating keys buffer, we overwrite all elements.
+//
+//    firstElemKey = getFirstElement(bucket)
+//    shouldOverwrite = blob.slot > bytes_to_slot(firstElemKey[8:16])
+//    if shouldOverwrite:
+// 	    for existingKey := seek prefix bytes(slot_to_rotating_buffer(blob.slot))
+//        bucket.delete(existingKey)
+//    bucket.put(key, blob)
+func (s *Store) SaveBlobsSidecar(ctx context.Context, blobSidecar *ethpb.BlobsSidecar) error {
 	ctx, span := trace.StartSpan(ctx, "BeaconDB.SaveBlobsSidecar")
 	defer span.End()
 	return s.db.Update(func(tx *bolt.Tx) error {
-		blobKey := blob.BeaconBlockRoot
-		insertTime := time.Now().Format(time.RFC3339)
-		ageBkt := tx.Bucket(blobsAgesBucket)
-		if err := ageBkt.Put(blobKey, []byte(insertTime)); err != nil {
+		encodedBlobSidecar, err := encode(ctx, blobSidecar)
+		if err != nil {
 			return err
 		}
-
 		bkt := tx.Bucket(blobsBucket)
-		enc, err := encode(ctx, blob)
-		if err != nil {
-			return err
+		c := bkt.Cursor()
+		key := blobSidecarKey(blobSidecar)
+		rotatingBufferPrefix := key[0:8]
+		var firstElementKey []byte
+		for k, _ := c.Seek(rotatingBufferPrefix); bytes.HasPrefix(k, rotatingBufferPrefix); k, _ = c.Next() {
+			if len(k) != 0 {
+				firstElementKey = k
+				break
+			}
+		}
+		// If there is no element stored at blob.slot % MAX_SLOTS_TO_PERSIST_BLOBS, then we simply
+		// store the blob by key and exit early.
+		if len(firstElementKey) == 0 {
+			return bkt.Put(key, encodedBlobSidecar)
+		} else if len(firstElementKey) != len(key) {
+			return fmt.Errorf(
+				"key length %d (%#x) != existing key length %d (%#x)",
+				len(key),
+				key,
+				len(firstElementKey),
+				firstElementKey,
+			)
 		}
-		return bkt.Put(blobKey, enc)
+		slotOfFirstElement := firstElementKey[8:16]
+		// If we should overwrite old blobs at the spot in the rotating buffer, we clear at data at that spot.
+		shouldOverwrite := blobSidecar.BeaconBlockSlot > bytesutil.BytesToSlotBigEndian(slotOfFirstElement)
+		if shouldOverwrite {
+			for k, _ := c.Seek(rotatingBufferPrefix); bytes.HasPrefix(k, rotatingBufferPrefix); k, _ = c.Next() {
+				if err := bkt.Delete(k); err != nil {
+					return err
```

Comment:
Aren't we possibly leaving the db in an inconsistent state?

Reply:
Yeah good point, might instead collect the errors